scio-core/src/main/scala/com/spotify/scio/values/BloomFilter.scala

```diff
@@ -0,0 +1,202 @@
+   * from which this [[BloomFilter]] was built.
+   */
+  val expectedFpp: Double = internal.expectedFpp()
+
+  /**
+   * Returns an estimate for the total number of distinct elements that have been added to this
+   * [[BloomFilter]]. This approximation is reasonably accurate if it does not exceed the value of
+   * `expectedInsertions` that was used when constructing the filter.
+   */
+  val approximateElementCount: Long = internal.approximateElementCount()
+
+  /**
+   * Check approximate presence of an element in the BloomFilter
+   *
+   * @return true if the element may be present, false if it is definitely not present.
+   */
+  def mayBeContains(t: T): Boolean = internal.mightContain(t)
+
+  // Java Serialization
+  private def writeObject(out: ObjectOutputStream): Unit = {
+    out.writeObject(funnel)
+    internal.writeTo(out)
+  }
+
+  private def readObject(in: ObjectInputStream): Unit = {
+    val funnel = in.readObject().asInstanceOf[Funnel[T]]
+    val internal = gBloomFilter.readFrom(in, funnel)
+    BloomFilter.setField("funnel", funnel)
+    BloomFilter.setField("internal", internal)
+  }
+
+  /**
+   * Add an element to this [[BloomFilter]]. It creates a copy of the underlying
+   * structure.
+   *
+   * For creating BloomFilters from large collections, use [[BloomFilter#apply(Iterable)]]
+   * instead.
+   */
+  @experimental
+  def put(t: T): BloomFilter[T] = {
+    val copyOfInternal = internal.copy()
+    copyOfInternal.put(t)
+    this.copy(internal = copyOfInternal)
+  }
+
+  /**
+   * Add an Itearable of elements to this [[BloomFilter]]. It creates a copy of the underlying
+   * structure.
+   *
+   * For creating BloomFilters from large collections, use [[BloomFilter#apply(Iterable)]]
+   * instead.
+   */
+  @experimental
+  def putAll(elements: Iterable[T]): BloomFilter[T] = {
+    val copyOfInternal = internal.copy()
+    val it = elements.iterator
+    while (it.hasNext) {
+      copyOfInternal.put(it.next())
+    }
+    this.copy(internal = copyOfInternal)
+  }
+}
+
+object BloomFilter extends ApproxFilterCompanion {
+
+  /**
+   * Constructor for [[BloomFilter]]
+   *
+   * @param iterable An iterable of elements to be stored in the BloomFilter
+   * @param fpProb allowed false positive probability
+   * @param f [[Funnel]] for the type [[T]]
+   * @return A [[BloomFilter]] for the given [[Iterable]] of elements.
+   */
+  def apply[T](
+    iterable: Iterable[T],
+    fpProb: Double
+  )(
+    implicit f: Funnel[T]
+  ): BloomFilter[T] = {
+    val numElements = iterable.size
+    val settings = BloomFilter.optimalBFSettings(numElements, fpProb)
+    require(
+      settings.numBFs == 1,
+      s"BloomFilter overflow: $numElements elements found, max allowed: ${settings.capacity}"
+    )
+
+    val bf = gBloomFilter.create[T](f, numElements, fpProb)
+    val it = iterable.iterator
+    while (it.hasNext) {
+      bf.put(it.next())
+    }
+    BloomFilter(bf, f)
```

Comment:
If we require `expectedInsertions`, we can check `expectedFpp` vs `fpProb` here and warn.

Reply:
Warn as in?

I am relying on the following failure checks from Guava
```
    checkArgument(
        expectedInsertions >= 0, "Expected insertions (%s) must be >= 0", expectedInsertions);
    checkArgument(fpp > 0.0, "False positive probability (%s) must be > 0.0", fpp);
    checkArgument(fpp < 1.0, "False positive probability (%s) must be < 1.0", fpp);
    checkArgument(numHashFunctions > 0, "numHashFunctions (%s) must be > 0", numHashFunctions);
    checkArgument(
        numHashFunctions <= 255, "numHashFunctions (%s) must be <= 255", numHashFunctions);
```

Reply:
```
var i = 0
while (it.hasNext) {
  i += 1
  // ...
}
// actual numbers exceeding user specified.
if (i >= expectedInsertions) // WARN
if (bf.expectedFpp > fpp) // WARN
```